hare/algorithm.go

```diff
@@ -88,140 +101,418 @@ func (proc *ConsensusProcess) createInbox(size uint32) chan *pb.HareMessage {
 		case msg := <-proc.inbox: // msg event
 			proc.handleMessage(msg)
 		case <-ticker.C: // next round event
-			proc.nextRound()
+			proc.onRoundEnd()
+			proc.advanceToNextRound()
+			proc.onRoundBegin()
 		case <-proc.CloseChannel(): // close event
-			log.Info("Stop event loop, instance aborted")
+			log.Info("Stop event loop, terminating")
 			return
 		}
+
+		if proc.terminating {
+			return
+		}
+	}
+}
+
+// verifies the message is contextually valid
+func (proc *ConsensusProcess) isContextuallyValid(m *pb.HareMessage) bool {
+	isSameIteration := iterationFromCounter(proc.k) != iterationFromCounter(m.Message.K)
+	currentRound := proc.currentRound()
+	switch MessageType(m.Message.Type) {
+	case PreRound:
+		return true
+	case Status:
+		return isSameIteration && currentRound == Round1
+	case Proposal:
+		return isSameIteration && currentRound == Round2 || currentRound == Round3
+	case Commit:
+		return isSameIteration && currentRound == Round3
+	case Notify:
+		return true
+	default:
+		log.Error("Unknown message type encountered during contextual validation: ", m.Message.Type)
+		return false
+	}
+}
+
+// verifies the message is syntactically valid
+func isSyntacticallyValid(m *pb.HareMessage) bool {
+	claimedRound := m.Message.K % 4
+
+	switch MessageType(m.Message.Type) {
+	case PreRound:
+		return true
+	case Status:
+		return claimedRound == Round1
+	case Proposal:
+		return claimedRound == Round2
+	case Commit:
+		return claimedRound == Round3
+	case Notify:
+		return true
+	}
+
+	log.Error("Unknown message type encountered during syntactic validation: ", m.Message.Type)
+	return false
+}
+
+func (proc *ConsensusProcess) validateCertificate(m *pb.HareMessage) bool {
+	if m == nil {
+		return false
+	}
+
+	if m.Cert == nil {
+		return false
+	}
+
+	if m.Cert.AggMsgs == nil {
+		return false
+	}
+
+	msgs := m.Cert.AggMsgs.Messages
+	if msgs == nil {
+		return false
+	}
+
+	if len(msgs) == 0 {
+		return false
+	}
+
+	// validate type, k and set
+	k := msgs[0].Message.K
+	s := NewSet(m.Message.Values)
+	for _, msg := range msgs {
+		// validate commit type
+		if MessageType(msg.Message.Type) != Commit {
+			return false
+		}
+
+		// validate same k
+		if msg.Message.K != k {
+			return false
+		}
+
+		// validate same set
+		g := NewSet(msg.Message.Values)
+		if !s.Equals(g) {
+			return false
+		}
+	}
+
+	// TODO: validate agg sig
+
+	return true
+}
+
+func roleFromRoundCounter(k int32) Role {
+	switch k % 4 {
+	case Round2:
+		return Leader
+	case Round4:
+		return Passive
+	default:
+		return Active
 	}
 }
 
 func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
-	// Note: layer is already verified by the broker
-	// Currently we reject past/future messages
+	// Note: instanceId is already verified by the broker
 
-	// verify iteration
-	if m.Message.K != proc.k {
-		log.Warning("Iteration mismatch. Message iteration is: %d expected: %d", m.Message.K, proc.k)
+	// validate context
+	if !proc.isContextuallyValid(m) {
```

Comment:
you should first check the syntactic validity and then the contextual since in when checking the contextual validity you are assuming that the message is syntactically valid

Reply:
Fixed.